x-pack/functionbeat/manager/gcp/cli_manager.go

```diff
@@ -0,0 +1,169 @@
+	if err != nil {
+		return err
+	}
+
+	c.log.Debugf("Successfully created function '%s'", name)
+	return nil
+}
+
+// Update updates the function.
+func (c *CLIManager) Update(name string) error {
+	c.log.Debugf("Starting updating function '%s'", name)
+	defer c.log.Debugf("Update complete for function '%s'", name)
+
+	update := true
+	err := c.deploy(update, name)
+	if err != nil {
+		return err
+	}
+
+	c.log.Debugf("Successfully updated function: '%s'", name)
+	return nil
+}
+
+// deploy uploads to bucket and creates a function on GCP.
+func (c *CLIManager) deploy(update bool, name string) error {
+	functionData, err := c.templateBuilder.execute(name)
+	if err != nil {
+		return err
+	}
+
+	executer := executor.NewExecutor(c.log)
+	executer.Add(newOpEnsureBucket(c.log, c.config))
+	executer.Add(newOpUploadToBucket(c.log, c.config, name, functionData.raw))
+
+	if update {
+		executer.Add(newOpUpdateFunction(c.log, c.tokenSrc, functionData.function.Name, functionData.function))
+	} else {
+		executer.Add(newOpCreateFunction(c.log, c.tokenSrc, c.location, name, functionData.function))
+	}
+
+	executer.Add(newOpWaitForFunction(c.log, c.tokenSrc))
+
+	ctx := &functionContext{}
+	if err := executer.Execute(ctx); err != nil {
+		if rollbackErr := executer.Rollback(ctx); rollbackErr != nil {
+			return errors.Wrapf(err, "could not rollback, error: %s", rollbackErr)
+		}
+		return err
+	}
+	return nil
+}
+
+// Remove removes a stack and unregister any resources created.
+func (c *CLIManager) Remove(name string) error {
+	c.log.Debugf("Removing function: %s", name)
+	defer c.log.Debugf("Removal of function '%s' complete", name)
+
+	functionData, err := c.templateBuilder.execute(name)
+	if err != nil {
+		return err
+	}
+
+	executer := executor.NewExecutor(c.log)
+	executer.Add(newOpDeleteFunction(c.log, c.location, functionData.function.Name, c.tokenSrc))
+	executer.Add(newOpDeleteFromBucket(c.log, c.config, name))
+
+	ctx := &functionContext{}
+	if err := executer.Execute(ctx); err != nil {
+		if rollbackErr := executer.Rollback(ctx); rollbackErr != nil {
+			return errors.Wrapf(err, "could not rollback, error: %s", rollbackErr)
+		}
+		return err
+	}
+
+	c.log.Debugf("Successfully deleted function: '%s'", name)
+	return nil
+}
+
+// Export prints the exported function data.
+func (c *CLIManager) Export(name string) error {
+	tmpl, err := c.templateBuilder.RawTemplate(name)
+	if err != nil {
+		return err
+	}
+	fmt.Println(tmpl)
+
+	return nil
+}
+
+// NewCLI returns the interface to manage function on Amazon lambda.
+func NewCLI(
+	log *logp.Logger,
+	cfg *common.Config,
+	provider provider.Provider,
+) (provider.CLIManager, error) {
+	config := &Config{}
+	if err := cfg.Unpack(config); err != nil {
+		return nil, err
+	}
+
+	builder, err := provider.TemplateBuilder()
+	if err != nil {
+		return nil, err
+	}
+
+	templateBuilder, ok := builder.(*defaultTemplateBuilder)
+	if !ok {
+		return nil, fmt.Errorf("not defaultTemplateBuilder")
+	}
```

Comment:
Does this indicate a programmer error, user error or internal bug?

What is the problem with the TemplateBuilder interface that we have to cast here?

Reply:
It's a programmer error. 

TemplateBuilder has an unexported function named `execute` for both providers with different return values. That function generates the actual templates, which is required for deployment and to exporting function templates. I haven't found any meaningful way to extract it from the template builder, so it needs to be cast to the actual type.

Reply:
I see. This makes me wonder if (in the future) the provider should construct the CLIManager.